From a738b5cbe940e96378945cee6764b76ba9cbcac6 Mon Sep 17 00:00:00 2001
From: Andriy Tkachuk <andriy_tkachuk@xyratex.com>
Date: Wed, 11 Dec 2013 10:27:00 +0200
Subject: [PATCH 1/7] rpc: update slot->sl_last_sent at m0_rpc_item_fini()

If the item finalized was the last sent one,
slot->sl_last_sent should be updated and point
to the previous element in the slot items list.

Reviewed-by: Nikita Danilov <nikita_danilov@xyratex.com>
---
 rpc/item.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/rpc/item.c b/rpc/item.c
index 4933b87..50b82c2 100644
--- a/rpc/item.c
+++ b/rpc/item.c
@@ -424,6 +424,7 @@ M0_EXPORTED(m0_rpc_item_init);
 void m0_rpc_item_fini(struct m0_rpc_item *item)
 {
 	struct m0_rpc_slot_ref *sref = &item->ri_slot_refs[0];
+	struct m0_rpc_slot *slot = sref->sr_slot;
 
 	M0_ENTRY("item: %p", item);
 
@@ -436,8 +437,13 @@ void m0_rpc_item_fini(struct m0_rpc_item *item)
 		m0_rpc_item_put(item->ri_reply);
 		item->ri_reply = NULL;
 	}
-	if (slot_item_tlink_is_in(item))
+	if (slot_item_tlink_is_in(item)) {
+		if (item == slot->sl_last_sent) {
+			slot->sl_last_sent =
+				slot_item_tlist_prev(&slot->sl_item_list, item);
+		}
 		slot_item_tlist_del(item);
+	}
 	if (itemq_tlink_is_in(item))
 		m0_rpc_frm_remove_item(item->ri_frm, item);
 	sref->sr_ow = invalid_slot_ref;
-- 
1.8.3.2

